src/coreclr/gc/gc.cpp

```diff
@@ -11541,6 +11532,477 @@ void gc_heap::rearrange_heap_segments(BOOL compacting)
+    size_free_regions += from->size_free_regions;
+    size_committed_in_free_regions += from->size_committed_in_free_regions;
+
+    from->num_free_regions_removed += from->num_free_regions;
+    from->num_free_regions = 0;
+    from->size_free_regions = 0;
+    from->size_committed_in_free_regions = 0;
+
+    from->head_free_region = nullptr;
+    from->tail_free_region = nullptr;
+
+    assert(this->num_free_regions == (this->num_free_regions_added - this->num_free_regions_removed));
+    assert(from->num_free_regions == (from->num_free_regions_added - from->num_free_regions_removed));
+}
+
+size_t region_free_list::get_num_free_regions()
+{
+#ifdef _DEBUG
+    size_t actual_count = 0;
+    heap_segment* last_region = nullptr;
+    for (heap_segment* region = head_free_region; region != nullptr; region = heap_segment_next (region))
+    {
+        last_region = region;
+        actual_count++;
+    }
+    assert (num_free_regions == actual_count);
+    assert (last_region == tail_free_region);
+    heap_segment* first_region = nullptr;
+    for (heap_segment* region = tail_free_region; region != nullptr; region = heap_segment_prev_free_region (region))
+    {
+        first_region = region;
+        actual_count--;
+    }
+    assert (actual_count == 0);
+    assert (head_free_region == first_region);
+#endif //_DEBUG
+    return num_free_regions;
+}
+#endif //USE_REGIONS
+
+void gc_heap::distribute_free_regions()
+{
+#ifdef USE_REGIONS
+    const int kind_count = large_free_region + 1;
+
+    // first step: accumulate the number of free regions and the budget over all heaps
+    // and move huge regions to global free list
+    size_t total_num_free_regions[kind_count] = { 0, 0 };
+    size_t total_budget[kind_count] = { 0,  0 };
+
+#ifdef MULTIPLE_HEAPS
+    for (int i = 0; i < n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+#else //MULTIPLE_HEAPS
+    {
+        gc_heap* hp = pGenGCHeap;
+        // just to reduce the number of #ifdefs in the code below
+        const int i = 0;
+#endif //MULTIPLE_HEAPS
+        for (int kind = basic_free_region; kind < kind_count; kind++)
+        {
+            total_num_free_regions[kind] += hp->free_regions[kind].get_num_free_regions();
+        }
+
+        global_free_huge_regions.transfer_regions (&hp->free_regions[huge_free_region]);
+
+        for (int gen = soh_gen0; gen < total_generation_count; gen++)
+        {
+            dynamic_data* dd_gen = hp->dynamic_data_of(gen);
+
+            // estimate how much we are going to need in gen - assume half the free list space gets used
+            ptrdiff_t new_allocation_gen = dd_new_allocation (dd_gen);
+            ptrdiff_t free_list_space_gen = generation_free_list_space (hp->generation_of(gen));
+            ptrdiff_t budget_gen = new_allocation_gen - (free_list_space_gen / 2);
+            dprintf (REGIONS_LOG, ("budget for gen %d on heap %d is %Id (new %Id, free %Id)", gen, i, budget_gen, new_allocation_gen, free_list_space_gen));
```

Comment:
this is the same logic we use in `decommit_ephemeral_segment_pages`, can we make a common method for it, something like `estimate_gen_growth`?

Reply:
Sure, we can do that...